File: frontend/oei-web/src/app/domain/model/institution/institution-invitation.ts
import { InstitutionRole } from './institution-role';

export type InstitutionInvitationStatus = 'PENDING' | 'ACCEPTED' | 'EXPIRED' | 'REVOKED';

export interface InstitutionInvitation {
  readonly id: string;
  readonly institutionId: string;
  readonly email: string;
  readonly role: InstitutionRole;
  readonly status: InstitutionInvitationStatus;
  readonly invitedBy: string;
  readonly invitedAt: string;
  readonly expiresAt: string;
}

export function createInstitutionInvitation(fields: InstitutionInvitation): InstitutionInvitation {
  return Object.freeze({ ...fields });
}

File: backend/domain/shared/src/main/java/global/oei/domain/shared/institution/InstitutionInvitation.java
package global.oei.domain.shared.institution;

import java.time.Instant;
import java.util.Objects;

/**
 * An institution's invitation for an employee to join its team space.
 */
public record InstitutionInvitation(
        String id,
        InstitutionId institutionId,
        String email,
        InstitutionRole role,
        InstitutionInvitationStatus status,
        String invitedBy,
        Instant invitedAt,
        Instant expiresAt) {

    public InstitutionInvitation {
        Objects.requireNonNull(id, "id must not be null");
        Objects.requireNonNull(institutionId, "institutionId must not be null");
        Objects.requireNonNull(email, "email must not be null");
        Objects.requireNonNull(role, "role must not be null");
        Objects.requireNonNull(status, "status must not be null");
    }

    /**
     * @return a new instance with {@link #status()} set to {@link InstitutionInvitationStatus#REVOKED},
     *         or throws if the invitation is not currently {@link InstitutionInvitationStatus#PENDING}
     */
    public InstitutionInvitation revoke() {
        if (status != InstitutionInvitationStatus.PENDING) {
            throw new IllegalStateException("only a PENDING invitation can be revoked, was " + status);
        }
        return new InstitutionInvitation(id, institutionId, email, role, InstitutionInvitationStatus.REVOKED, invitedBy, invitedAt, expiresAt);
    }
}
